### PR TITLE
Added cache for configuration files

### LINKED ISSUE #195: Repository configuration cache
Artipie performs a lot of read operations for repository configs, it would be better to cache these configs in memory. The config can be updated at any time, so we need to drop cached value after some time, e.g. after few minutes. Consider using soft references to avoid OOM errors when working with many repos.

### PR DESCRIPTION
**Pull request by genryxy**
Closes #195 
Added cache with soft references with eviction which is equal to two minutes for repository configuration files and for aliases with configs of storage.

### CODE REVIEW

**Review by olenagerasimova — APPROVED** [review]
@genryxy thanks, looks good for me

**Comment by genryxy** [comment]
@g4s8 could you please check / merge

File: src/main/java/com/artipie/RepositoriesFromStorage.java [hunk 1]
```diff
@@ -4,19 +4,65 @@
...
+     */
+    private static LoadingCache<KeyAndStorage, Single<StorageAliases>> aliases;
+
+    static {
+        final CacheLoader<KeyAndStorage, Single<String>> ldrconfigs = new CacheLoader<>() {
+            @Override
+            public Single<String> load(final KeyAndStorage pair) {
+                return Single.fromFuture(
+                    new ConfigFile(pair.key)
+                        .valueFrom(pair.storage)
```
Comment (g4s8) [inline src/main/java/com/artipie/RepositoriesFromStorage.java:43]: I don't like private fields access here because it increases the coupling between components. I'd suggest introducing a method in `KeyAndStorage` which can access these private fields, e.g.
```
// class KeyAndStorage
CompletionStage<Content> content() {
  return new ConfigFile(this.key).valueFrom(this.storage);
}
```
Also, in this case, it could have a more meaningful name, e.g. `ConfigContent` or similar

File: src/test/java/com/artipie/RepositoriesFromStorageTest.java [hunk 2]
```diff
@@ -85,7 +85,7 @@ void throwsExceptionWhenConfigYamlAbsent() {
             this::repoConfig
         );
         MatcherAssert.assertThat(
-            result.getCause(),
+            result.getCause().getCause(),
```
Comment (g4s8) [inline src/test/java/com/artipie/RepositoriesFromStorageTest.java:88]: This change looks not related to PR modification, isn't it?

**Review by g4s8 — CHANGES_REQUESTED** [review]
@genryxy looks good, please check a few comments

File: src/test/java/com/artipie/RepositoriesFromStorageTest.java (hunk 2 above)
Comment (genryxy) [reply src/test/java/com/artipie/RepositoriesFromStorageTest.java:88]: @g4s8 now `ValueNotFoundException` is wrapped with `java.util.concurrent.ExecutionException` therefore this test was changed

File: src/main/java/com/artipie/ArtipieProperties.java [hunk 3]
```diff
@@ -54,6 +54,21 @@ public String version() {
...
+     * Obtains timeout of caching configuration files in milliseconds.
+     * @return Timeout of caching files with configuration.
+     */
+    public int configCacheTimeout() {
+        int time;
+        try {
+            time = Integer.parseInt(this.properties.getProperty("artipie.config.cache.timeout"));
+        } catch (final NumberFormatException exc) {
+            // @checkstyle MagicNumberCheck (1 line)
+            time = 2 * 60 * 1000;
```
[the remark this answers is not in the document]
Comment (genryxy) [reply src/main/java/com/artipie/ArtipieProperties.java]: @g4s8 I've replaced with `Integer.getInteger("artipie.config.cache.timeout", 2 * 60 * 1000)` although it does not throw any exception in case of mistyping in `artipie.properties` and I had to set property value at first

**Comment by genryxy** [comment]
@g4s8 thanks, corrected and answered

File: src/main/java/com/artipie/RepositoriesFromStorage.java [hunk 4]
```diff
@@ -35,33 +35,33 @@
...
+        System.setProperty(
+            ArtipieProperties.CONFIG_TIMEOUT,
+            new ArtipieProperties().configCacheTimeout()
+        );
+        final int timeout = Integer.getInteger(ArtipieProperties.CONFIG_TIMEOUT, 2 * 60 * 1000);
         RepositoriesFromStorage.configs = CacheBuilder.newBuilder()
             .expireAfterWrite(timeout, TimeUnit.MILLISECONDS)
-            .softValues().build(ldrconfigs);
+            .softValues()
+            .build(
```
Comment (g4s8) [inline src/main/java/com/artipie/RepositoriesFromStorage.java:46]: I mean can we use something like `build(config -> config.configContent())` here? Or `CacheLoader` is not a `FunctionalInterface`?

**Review by g4s8 — COMMENTED** [review]
@genryxy please check one more questions

File: src/main/java/com/artipie/RepositoriesFromStorage.java (hunk 4 above)
Comment (genryxy) [reply src/main/java/com/artipie/RepositoriesFromStorage.java:46]: @g4s8 got it. `CacheLoader` is an abstract class.

**Review by g4s8 — APPROVED** [review]
@genryxy thanks
